src/lib/lossless_cast.hpp

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include <optional>
+#include <type_traits>
+
+#include "all_type_variant.hpp"
+#include "null_value.hpp"
+
+/**
+ * This file defines the following casting functions that guarantee **lossless** conversion between data types and
+ * return std::nullopt if no lossless conversion was possible.
+ *      lossless_cast<Target>(Source)
+ *      lossless_variant_cast<Target>(AllTypeVariant)
+ *      lossless_variant_cast(Source, DataType)
+ *
+ * "Lossless" means just that: No information must be lost during casting. Typical lossful operations would be
+ *      - float-to-int casts with the float having a non-zero fractional part
+ *      - long-to-int casts where the long was > MAX_INT
+ *      - string-to-int where the string contained no number, or a number > MAX_INT.
+ *      - double-to-float where the significand of the double has bits set that cannot be represented in float
+ *
+ * Whenever possible you should use these casts over "comfortable" casts such as static_cast or boost::lexical_cast.
+ * Those might incur information loss and might thus introduce bugs. See #1306, #1525.
+ *
+ * THINK ABOUT AN ALGORITHM'S DATA, THE POSSIBLE STATES IT MIGHT HAVE, AND HANDLE THEM PROPERLY.
```

Comment:
This applies everywhere, doesn't it?

Reply:
Ah, sure... I guess I was just surprised by what I perceived as careless usage of the lossy type_cast all over the code base.  
Removed the comment.